addons/controllers/tanzuclusterbootstrap_controller.go

```diff
@@ -0,0 +1,586 @@
+	"github.com/vmware-tanzu/tanzu-framework/addons/pkg/util"
+	"github.com/vmware-tanzu/tanzu-framework/addons/predicates"
+	runtanzuv1alpha3 "github.com/vmware-tanzu/tanzu-framework/apis/run/v1alpha3"
+)
+
+// TanzuClusterBootstrapReconciler reconciles a TanzuClusterBootstrap object
+type TanzuClusterBootstrapReconciler struct {
+	client.Client
+	Log     logr.Logger
+	Scheme  *runtime.Scheme
+	context context.Context
+
+	// internal properties
+	controller    controller.Controller
+	dynamicClient dynamic.Interface
+	// on demand dynamic watches for provider refs
+	providerWatches map[string]client.Object
+	// discovery client for looking up api-resources and preferred versions
+	cachedDiscoveryClient discovery.CachedDiscoveryInterface
+	// cache for resolved api-resources so that look up is fast (cleared periodically)
+	providerGVR map[schema.GroupKind]*schema.GroupVersionResource
+}
+
+// NewtanzuClusterBootstrapReconciler returns a reconciler for TanzuBootstrapCluster
+func NewtanzuClusterBootstrapReconciler(c client.Client, log logr.Logger, scheme *runtime.Scheme) *TanzuClusterBootstrapReconciler {
+	return &TanzuClusterBootstrapReconciler{
+		Client: c,
+		Log:    log,
+		Scheme: scheme,
+	}
+}
+
+//+kubebuilder:rbac:groups=run.tanzu.vmware.com,resources=tanzuclusterbootstraps,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=run.tanzu.vmware.com,resources=tanzuclusterbootstraps/status,verbs=get;update;patch
+
+// SetupWithManager performs the setup actions for an TanzuBootstrapCluster controller, using the passed in mgr.
+func (r *TanzuClusterBootstrapReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
+	ctrlr, err := ctrl.NewControllerManagedBy(mgr).
+		For(&clusterapiv1beta1.Cluster{}).
+		Watches(
+			&source.Kind{Type: &runtanzuv1alpha3.TanzuKubernetesRelease{}},
+			handler.EnqueueRequestsFromMapFunc(r.TKRToClusters),
+			builder.WithPredicates(
+				predicates.TKR(r.Log),
+			),
+		).
+		Watches(
+			&source.Kind{Type: &runtanzuv1alpha3.TanzuClusterBootstrap{}},
+			handler.EnqueueRequestsFromMapFunc(r.TanzuClusterBootstrapToClusters),
+			builder.WithPredicates(
+				predicates.TKR(r.Log),
+			),
+		).
+		WithOptions(options).
+		WithEventFilter(clusterApiPredicates.ResourceNotPaused(r.Log)).
+		Build(r)
+	if err != nil {
+		r.Log.Error(err, "Error creating an addon controller")
+		return err
+	}
+
+	r.controller = ctrlr
+	r.context = ctx
+	dynClient, err := dynamic.NewForConfig(mgr.GetConfig())
+	if err != nil {
+		r.Log.Error(err, "Error creating dynamic client")
+		return err
+	}
+	r.dynamicClient = dynClient
+	r.providerWatches = make(map[string]client.Object)
+
+	r.providerGVR = make(map[schema.GroupKind]*schema.GroupVersionResource)
+	clientset := kubernetes.NewForConfigOrDie(mgr.GetConfig())
+	r.cachedDiscoveryClient = cacheddiscovery.NewMemCacheClient(clientset.Discovery())
+
+	go r.periodicGVRCachesClean()
```

Comment:
unrelated:  I was expecting client-go handles the cache invalidation to make sure data consistency.

Reply:
This is for calling Invalidate() on CachedDiscovery which has a helper for getting preferred apiResources that the server has, since getting all api-resources is an expensive call the cached client helps, with the caveat that it needs to be invalidated periodically if new CRD's are installed and need to be discovered.

Nothing to do with controller-runtime cache for getting specific instance of a resource.